lib/puppet/provider/snmp_user/nxapi.rb

```diff
@@ -0,0 +1,142 @@
+# The NXAPI provider for cisco snmp_user.
+#
+# November, 2014
+#
+# Copyright (c) 2014-2015 Cisco and/or its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'cisco_node_utils' if Puppet.features.cisco_node_utils?
+begin
+  require 'puppet_x/cisco/autogen'
+rescue LoadError # seen on master, not on agent
+  # See longstanding Puppet issues #4248, #7316, #14073, #14149, etc. Ugh.
+  require File.expand_path(File.join(File.dirname(__FILE__), '..', '..', '..',
+                                     'puppet_x', 'cisco', 'autogen.rb'))
+end
+
+Puppet::Type.type(:snmp_user).provide(:nxapi) do
+  desc 'The Cisco NXAPI provider for snmp_user.'
+
+  confine feature: :cisco_node_utils
+  defaultfor operatingsystem: :nexus
+
+  mk_resource_methods
+
+  REQUIRES_AUTH_PROPS = [:password, :privacy, :private_key, :localized_key, :engine_id, :auth]
+
+  def initialize(value={})
+    super(value)
+    @snmpuser = Cisco::SnmpUser.users[@property_hash[:name]]
+    @property_flush = {}
+    debug 'Created provider instance of snmp_user'
+  end
+
+  def self.properties_get(snmpuser_name, v)
+    debug "Checking instance, SnmpUser #{snmpuser_name}"
+
+    current_state = {
+      ensure:      :present,
+      name:        snmpuser_name,
+      engine_id:   v.engine_id,
+      roles:       v.groups,
+      auth:        v.auth_protocol,
+      password:    v.auth_password,
+      privacy:     v.priv_protocol,
+      private_key: v.priv_password,
+    }
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    snmpusers = []
+    Cisco::SnmpUser.users.each do |snmpuser_name, v|
+      snmpusers << properties_get(snmpuser_name.split(' ')[0], v)
```

Comment:
@hunsbea Well spotted bug. However, the issue isn't uniqueness per se. The issue lies in the fact that the provider will use the values obtained from the currently configured resource. I need to think of a way to only use the puppet-provided values and ignore the currently set items.

I don't believe it's valid to have 1 user configured twice on the system? This doesn't really make sense?